### PR TITLE
Use `deepmerge` package to merge the state instead of using spread syntax

### PR DESCRIPTION
**Pull request by coldlink**
## What does this change?
Following on from team conversations, using the spread syntax to modify the state works, but when merging object within objects (deep objects), as JS only merges at the top level by default, we have to spread the inner property too, this leads to nested spread syntax, as well as making it difficult to reason about what it being merged in, making it more prone to errors.

This PR introduces the [`deepmerge`](https://github.com/TehShrike/deepmerge) library, which is made to solve this exact problem when merging objects with deep enumerable properties. The API is very simple to use, and typescript is also able to reason about the properties of the objects being merged, which is very helpful in development. It also creates a new object by default rather than modifying an existing object, which means we don't have to worry about introducing mutability errors.

### CODE REVIEW

File: cypress/integration/sign_in.spec.js [hunk 1]
```diff
@@ -15,9 +15,7 @@ describe('Sign in flow', () => {
     });
 
     it('Has no detectable a11y violations on sign in page with error', () => {
-      cy.visit(
-        '/signin',
-      );
+      cy.visit('/signin');
```
Comment (coldlink) [inline cypress/integration/sign_in.spec.js:18]: prettier was complaining, so fixed this here

**Review by oliverlloyd — APPROVED** [review]
Personally, I am fine with using the spread operator. I think in the description you mention it mutates state but actually its the pattern you most often see in examples for how to not mutate things - not sure if that affects the choice here.

I'm generally a fan of keeping code as close to vanilla javascript as possible as this makes it easier to read and easier for new developers to onboard. The more abstractions you include the more work you're asking people who are unfamiliar with the code to do. We don't feel the weight of this work because we're the ones adding things step by step but it can be a big problem for new joiners, especially after a few years of similar changes...

That said, I get that this is a subjective area and my feelings aren't all that strong. It also helps that this abstraction is an established library with its own documentation available so the process of learning it is not so bad

Basically: 👍

**Comment by oliverlloyd** [comment]
> Personally, I am fine with using the spread operator

Thinking about this. I realised the only reason I'm fine with using the spread operator is because I have fought through the pain and hassle of making my brain understand it. I well remember how confusing it could be at the start when it was added to the spec!

**Comment by coldlink** [comment]
> > Personally, I am fine with using the spread operator
> 
> Thinking about this. I realised the only reason I'm fine with using the spread operator is because I have fought through the pain and hassle of making my brain understand it. I well remember how confusing it could be at the start when it was added to the spec!

yeah, i think in the Sign in with okta PR there is a case where something like the following happens:
```js
state = {
  ...state,
  pageData: {
    ...state.pageData,
    email,
  },
  globalMessage: {
    ...state.globalMessage,
    error: message
  }
}
```
and that's super confusing to reason what's going on there when you can just do
```js
deepmerge(state, {
  pageData: {
    email,
  },
  globalMessage: {
    error: message,
  },
})
```

File: src/server/routes/consents.ts [hunk 2]
```diff
@@ -518,22 +513,18 @@ router.get(
...
 
-      state = {
-        ...state,
+      state = deepmerge(state, {
         pageData: {
-          ...state.pageData,
           ...(await read(req.ip, sc_gu_u, state.pageData.geolocation)),
         },
-      };
+      } as RequestState);
```
[the remark this answers is not in the document]
Comment (coldlink) [reply src/server/routes/consents.ts:520]: yeah it's fine, it's similar to other cases in this PR where a `)` is added after using the `deepmerge` method

for some reason this is the only case I had to tell typescript to `as RequestState` as it was unable to infer the type, but in the other cases it worked fine

File: src/server/routes/signIn.ts [hunk 3]
```diff
@@ -54,18 +55,16 @@ router.post(
...
 
-      state = {
-        ...state,
-        globalMessage: {
-          ...state.globalMessage,
-          error: message,
-        },
-      };
-
+      // re-render the sign in page on error
```
[the remark this answers is not in the document]
Comment (coldlink) [reply src/server/routes/signIn.ts:58]: we're still updating the state here, i pass the `deepmerge` method directly into the `renderer`, i just added this comment to help explain what was happening

**Pull request merged**